keps/sig-network/3866-nftables-proxy/README.md

```diff
@@ -1463,22 +1439,10 @@ We expect no non-infra related flakes in the last month as a GA graduation crite
 
 #### Scalability & Performance tests
 
-```
-<<[UNRESOLVED perfscale ]>>
-
-- For the control plane side, the existing scalability tests are
-  probably reasonable, assuming we implement the same
-  `NetworkProgrammingLatency` metric as the existing backends.
-
-- For the data plane side, there are tests of
-  `InClusterNetworkLatency`, but no one is really looking at the
-  results yet and they may need work before they are useable.
+We have an [nftables scalability job]. Initial performance is fine; we
+have not done a lot of further testing/improvement yet.
```

Comment:
But we do have evidence that it is more performant and should scale fine? I suspect this undersells it a bit :-)

Reply:
For GA we should probably take a look at how the 5,000 node jobs do.

Reply:
We do have some numbers but I don't have them handy and wanted to push this sooner rather than later...

> For GA we should probably take a look at how the 5,000 node jobs do.

Kube-proxy performance does not directly scale in any way with node count. If the 5000 node job creates more services than the 100 node job, then _that_ affects the metrics, but you don't actually need a large cluster to test large numbers of services; from _kube-proxy's_ perspective, having 10,000 Services that all point to the same 3 Pods is exactly the same as having 10,000 Services that point to 30,000 unique pods.

@npinaeva has been doing some tests of that sort, and at the moment nftables kube-proxy's performance with many many services is much closer to old pre-partial-sync iptables kube-proxy's performance than current iptables kube-proxy's performance, but well, we were waiting to have better perf numbers before we got to heavily into optimization.

Reply:
What dimension of performance are we talking here, network programming latency is the main pain point today,  dataplane performance as the linear search evaluation was always the main complain, but it seems that is more a synthetic problem

Reply:
She was testing dataplane performance but ran into the programming latency performance problem while trying to test that. It's not clear yet exactly what aspect of the test data was making it slow, other than that the slowness was in the actual reload operation in the kernel, not in the userspace parsing, etc. (In theory, nft should not be slower than iptables, given that modern iptables is just translating to nftables underneath. Though it's possible that, e.g., adding an element to a set is unexpectedly slow, which is something that would affect our nftables performance but not our iptables-via-iptables-nft performance.)

For dataplane performance there wasn't any observable improvement at the size she was testing, which is kind of expected because we hadn't ever had any complaints at the sizes kube-proxy could already support well. (In particular, pre-`MinimizeIPTablesRestore`, it was basically impossible to have so many services that the linear search in the `nat` table would affect dataplane performance.)

In the `filter` table, the limits are much smaller, and we've hit them twice (https://github.com/kubernetes/kubernetes/pull/56164 / https://github.com/kubernetes/kubernetes/pull/95252), but it's not a "things gradually get slower" problem, it's an "everything works fine until suddenly the entire network is completely broken" problem. We should try to reproduce that and confirm that nftables doesn't hit the same problem.